### PR TITLE
Migrate check and get filter functions from update_filter_d.sh to util.sh

### PR DESCRIPTION
**Pull request by stefanos1316**
It is necessary to migrate some of the update_filter_d.sh functionalities inside the util.sh, because wps_scanner_d will use them too.

### CODE REVIEW

File: epidose/device/wps_scanner_d.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#!/bin/sh
+#
+# Check for available networks and connect to the closest one using the WPS PBS
+#
+# Copyright 2020 Diomidis Spinellis
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Your name should be here, right?

File: epidose/device/wps_scanner_d.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    exit_code=$?
+    
+    # Obtain dynamic IP address
+    dhclient wlan0
+  fi
+
+  # Turn off the orange blinking of the WPS LED
+  #run_python device_io --orange-off
+  
+  return $exit_code
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: `exit_code` will not have a value if the `if` is not executed.

File: epidose/device/wps_scanner_d.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+UTIL="$(dirname "$0")/util.sh"
+
+# Source common functionality (logging, WiFi)
+# shellcheck source=epidose/device/util.sh
+SERVER_URL=none . "$UTIL"
+
+wps_connect()
+{
+  # Turn on the orange blinking of the WPS LED
+  #run_python device_io --orange-on
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: You also need to 
* Wait for the button to be pressed 
* Indicate whether the connection succeeded or not (e.g. turn on Green or Red)
* Turn off the LED when done after a small pause.
* Keep in mind that the *watchdog* will also be flashing the LED.  Consequently, you need to inform the watchdog to pause its LED activities, perhaps by creating (and then removing) a file in `/var/lock/`.

File: epidose/device/supervisord.conf [hunk 4]
```diff
@@ -68,5 +68,19 @@ redirect_stdout=true           ; redirect proc stderr to stdout (default false)
...
+stopasgroup=true               ; send stop signal to the UNIX process group (default false)
+killasgroup=true               ; SIGKILL the UNIX process group (def false)
+user=root                      ; setuid to this UNIX account to run the program
+redirect_stdout=true           ; redirect proc stderr to stdout (default false)
+stderr_logfile=/var/log/%(program_name)s ; stderr log path, NONE for none; default AUTO
+stderr_logfile_maxbytes=1MB    ; max # logfile bytes b4 rotation (default 50MB)
+
 [group:epidose]
-programs=beacon_tx,beacon_rx,update_filter,upload_seeds,watchdog
+programs=beacon_tx,beacon_rx,update_filter,upload_seeds,wps_scanner,watchdog
```
Comment (dspinellis) [inline epidose/device/supervisord.conf]: Please order programs alphabetically

File: epidose/device/wps_scanner_d.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+# Pick up utility functions relative to the script's source code
+UTIL="$(dirname "$0")/util.sh"
+
+# Source common functionality (logging, WiFi)
+# shellcheck source=epidose/device/util.sh
+SERVER_URL=none . "$UTIL"
+
+wps_connect()
+{
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Don't you need to power-on the WiFi and lock the access to it?  See the script that updates the Cuckoo filter.

File: Makefile [hunk 6]
```diff
@@ -27,5 +27,6 @@ fast-install:
 	cp epidose/device/upload_seeds_d.sh \
 		epidose/device/update_filter_d.sh \
 		epidose/device/shutdown_epidose.sh \
+		epidose/device/wps_scanner_d.sh \
 		epidose/device/util.sh $(OPT)/bin/
```
Comment (dspinellis) [inline Makefile]: Please order programs alphabetically (initially my mistake here).

File: epidose/device/wps_scanner_d.sh [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+  # Turn off the orange blinking of the WPS LED
+  #run_python device_io --orange-off
+  
+  return $exit_code
+}
+
+# Wait for WPS button to be pressed
+while : ; do
+  while ! wps_connect; do
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: I don't think you need a second loop.  Use `if`?

File: epidose/device/wps_scanner_d.sh [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	log "Failed to connect using WPS PBS"
+  done
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: After obtaining WiFi access you need to turn it off to conserve energy.  Also, consider refreshing the Cuckoo filter, if needed.  See the Cuckoo filter download script.

**Review by dspinellis — CHANGES_REQUESTED** [review]
Overall in the right direction; well done! I liked the many comments you added. Some changes are required.

File: epidose/device/wps_scanner_d.sh (hunk 1 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Yes, I will replace it.

File: epidose/device/wps_scanner_d.sh (hunk 2 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Then I will add the exit_code in the if loop.

File: epidose/device/wps_scanner_d.sh (hunk 3 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Regarding the first bullet point, I guess using the `run_python device_io --wifi-wait` will do the job, right?
About the last one, I guess you mean by using the `touch /var/lock/watchdog_d` will pause the execution of the corresponding daemon, right?
But, I do not see the watchdog daemon inside the setup.py or the Makefile.
Also, I cannot find it among my processes `$ ps -aux | grep epidose`.

File: epidose/device/supervisord.conf (hunk 4 above)
Comment (stefanos1316) [reply epidose/device/supervisord.conf]: Done!

File: epidose/device/wps_scanner_d.sh (hunk 5 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Ok, done!

File: Makefile (hunk 6 above)
Comment (stefanos1316) [reply Makefile]: Ok, done!

File: epidose/device/wps_scanner_d.sh (hunk 7 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: True. Replaced.

File: epidose/device/wps_scanner_d.sh (hunk 8 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Ok, I guess I can add a check on the filter to check if it is greater than 6 hours and act accordingly before releasing the WiFi, yes?

Comment (dspinellis) [reply epidose/device/wps_scanner_d.sh]: DRY (don't repeat yourself). The check is already in the Cyckoo filter update daemon.  So think of a design that will allow the two to work together.

File: epidose/device/wps_scanner_d.sh (hunk 3 above)
Comment (dspinellis) [reply epidose/device/wps_scanner_d.sh]: 1. Yes
2. Yes, something that watchdog can monitor
3. Watchdog is running through *supervisor* and `setup.py`
```
$ sudo supervisorctl status
epidose:beacon_rx                RUNNING   pid 446, uptime 5 days, 5:58:33
epidose:beacon_tx                RUNNING   pid 19546, uptime 0:09:35
epidose:update_filter            RUNNING   pid 447, uptime 5 days, 5:58:33
epidose:upload_seeds             RUNNING   pid 448, uptime 5 days, 5:58:33
epidose:watchdog                 RUNNING   pid 450, uptime 5 days, 5:58:32
```

Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: This is what I get from watchdog's log:

`Traceback (most recent call last):
  File "/opt/venvs/epidose/lib/python3.7/site-packages/epidose/device/watchdog_d.py", line 154, in <module>
    main()
  File "/opt/venvs/epidose/lib/python3.7/site-packages/epidose/device/watchdog_d.py", line 142, in main
    v = get_battery_voltage()
  File "/opt/venvs/epidose/lib/python3.7/site-packages/epidose/device/device_io.py", line 156, in get_battery_voltage
    spi = spi_setup()
  File "/opt/venvs/epidose/lib/python3.7/site-packages/epidose/device/device_io.py", line 137, in spi_setup
    spi.open(0, 0)
FileNotFoundError: [Errno 2] No such file or directory`
```

(venv) pi@raspberrypi:~/epidose $ sudo supervisorctl status
epidose:beacon_rx                RUNNING   pid 22995, uptime 3:19:21
epidose:beacon_tx                RUNNING   pid 24070, uptime 0:02:15
epidose:update_filter            RUNNING   pid 22996, uptime 3:19:21
epidose:upload_seeds             RUNNING   pid 22997, uptime 3:19:21
epidose:watchdog                 FATAL     Exited too quickly (process log may have details)

```
I guess when I will have the actual device this will be solved.

Comment (dspinellis) [reply epidose/device/wps_scanner_d.sh]: You're not running it on a real device, right?  Where is your Epidose package? Can you please coordinate to track it?

File: epidose/device/wps_scanner_d.sh (hunk 8 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Then should I add the cuckoo filter check in the util.sh, as it is needed from more than one script, and then invoke it in the update_filter_d and wps_scanner_d?

File: epidose/device/wps_scanner_d.sh (hunk 3 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: I am tracking it every day. Its latest location it is the Athens Airmail Center.
I can see something similar in the Hungarian's postal tracking service.

**Comment by dspinellis** [comment]
Much improved!  Only a few issues remaining, plus testing on the actual hardware.

File: epidose/device/wps_scanner_d.sh (hunk 3 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: I run a couple of tests in order to see if the LED lights are flashing as expected.
Seems the `touch /var/lock/suspend-watchdog-led` does not stop the blinking status light.
I have also tried using the `supervisorctl stop epidose:watchdog` but the status light is still blinking.

Comment (dspinellis) [reply epidose/device/wps_scanner_d.sh]: You can enable/add debugging to the watchdog in order to see what is going on.

**Comment by stefanos1316** [comment]
Do you require any more changes for this, or should I squash them in a single commit?

**Comment by dspinellis** [comment]
Thanks!  Please change the documentation of postconditions that return a value, to simply indicate a return value (this is not typically defined as a postcondition).  Afterwards, please test and squash for another review.

**Comment by stefanos1316** [comment]
Ok, I will change and rebase again.

**Comment by dspinellis** [comment]
I am confused.  These are two (not one) very small commits, that don't seem to implement what the PR indicates.  Can you please clarify, indicating which commit SHA I should review?

**Comment by stefanos1316** [comment]
When I squashed the commits, I mistakenly squashed them into this [commit](https://github.com/eellak/epidose/pull/41/commits/dceea84ac5fb88e054b156d8390f2eaceeea6ec1), which is a commit regarding the "Mofify fake-hwclock to use RTC". I am not sure how to resolve this.

**Comment by dspinellis** [comment]
https://stackoverflow.com/questions/134882/undoing-a-git-rebase

**Comment by stefanos1316** [comment]
Please let me know if everything is fine now.

File: epidose/device/update_filter_d.sh [hunk 9]
```diff
@@ -39,21 +39,13 @@ UTIL="$(dirname "$0")/util.sh"
 # Wait until a filter file is required
 wait_till_filter_needed()
 {
-  if [ -r "$FILTER" ] ; then
```
Comment (dspinellis) [inline epidose/device/update_filter_d.sh:42]: How are these changes related to WPS?  Please include in this commit only the WPS changes.

File: epidose/device/update_filter_d.sh [hunk 10]
```diff
@@ -39,21 +39,13 @@ UTIL="$(dirname "$0")/util.sh"
...
-      log "Sleeping for $to_sleep s"
-      sleep $to_sleep
-      log "Waking up from sleep; new filter is now required"
-    fi
-    log "Filter's age ($filter_age s) makes it stale; update required"
-  else
-    log "No filter available; download required"
-    mkdir -p "$(dirname $FILTER)"
+  # Calculates whether the cuckoo filter is stale or fresh
+  # and return a value accordingly
```
Comment (dspinellis) [inline epidose/device/update_filter_d.sh]: What value?  How is it returned? (These fixes apply to a separate PR, which I presume will deal with the cuckoo filter.)

File: epidose/device/wps_scanner_d.sh [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  run_python device_io --wifi-wait
+  wifi_acquire
+
+  network_status=$(ifconfig wlan0 | grep "inet ")
+
+  # Check if device is connected to any network
+  if [ -z "$network_status" ] ; then
+    log "Iniating WPS connection"
+    
+    # Connected to the network that supports WPS and has the lowest signal level
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: You mean highest, right?

File: epidose/device/wps_scanner_d.sh [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    wpa_cli scan_results | grep WPS | sort -r -k3 | tail -1 | awk '{print $1}' | xargs wpa_cli -i wlan0 wps_pbc
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Impressive pipeline.  What happens if no WPS network is found?  Have you tested it?  Do you need some special handling?

File: epidose/device/wps_scanner_d.sh [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    log "Trying to connect to $(wpa_cli scan_results | grep WPS | sort -r -k3 | tail -1 | awk '{print $NF}')"
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Store this in a variable, so that it's always the same and not recomputed.

File: epidose/device/wps_scanner_d.sh [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    
+    # Obtain dynamic IP address
+    dhclient wlan0
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Won't the script exit if dhclient fails?  This about this failure mode for all commands.

File: epidose/device/wps_scanner_d.sh [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    log "Failed to connect after using the WPS PBS"
+  else
+    check_cuckoo_filter
+    if [ "$to_sleep" -eq 0 ] ; then
+      log "Updating filter"
+      try_to_get_filter_from_server
+    fi
+    #run_python check_interface_risk "$FILTER" || :
+  fi
+  wifi_release
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Try to place wifi_acquire and release in symmetric positions, making it easy to verify that these calls always match.

File: epidose/device/wps_scanner_d.sh [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    wpa_cli scan_results | grep WPS | sort -r -k3 | tail -1 | awk '{print $1}' | xargs wpa_cli -i wlan0 wps_pbc
+    log "Trying to connect to $(wpa_cli scan_results | grep WPS | sort -r -k3 | tail -1 | awk '{print $NF}')"
+    
+    # Obtain dynamic IP address
+    dhclient wlan0
+
+    # Suspend watchdog's LED activities
+    touch "$SUSPEND_WATCHDOG_LED"
+
+    check_network_again=$(ifconfig wlan0 | grep "inet ")
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: But you are not checking network again.  Consider a more appropriate name.

**Review by dspinellis — CHANGES_REQUESTED** [review]
The WPS part seems close enough; well done!  A number of changes are however required.

File: epidose/device/update_filter_d.sh (hunk 9 above)
Comment (stefanos1316) [reply epidose/device/update_filter_d.sh:42]: WPS requires some functions found inside the update_filter_d script (such as check the age of cuckoo filter and get a new filter from the server). Therefore, I had to think of a design to help update_filter_d and wps_scanner_d to work together. This is why I migrated some functionalities of the update_filter_d inside the util.sh.

File: epidose/device/wps_scanner_d.sh (hunk 11 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Yes, highest.

File: epidose/device/update_filter_d.sh (hunk 10 above)
Comment (stefanos1316) [reply epidose/device/update_filter_d.sh]: So should I create a PR where I migrate the checks found inside the update_filter_d to the util.sh?
We need this because some of the functions found inside update_filter_d are required by the wps_scanner_d.

File: epidose/device/wps_scanner_d.sh (hunk 13 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Ok

File: epidose/device/wps_scanner_d.sh (hunk 12 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: No, I haven't tested that. I will check the possible return values and think of a solution.

File: epidose/device/wps_scanner_d.sh (hunk 15 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Ok, now the wifi_acquire is done before the wps_connect function call and the wifi_release after the function.

File: epidose/device/wps_scanner_d.sh (hunk 12 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Here is a small test.
The `wpa_cli scan_results` will return the following in no WPS network is found:

```
pi@raspberrypi:~/epidose $ wpa_cli scan_results
Selected interface 'p2p-dev-wlan0'
bssid / frequency / signal level / flags / ssid
pi@raspberrypi:~/epidose $ ifconfig
```

So I can write a check and if the $(wpa_cli scan_results | grep WPS) returns an empty string
then we should blink the red light, if not then we try to connect to the network, right?

File: epidose/device/update_filter_d.sh (hunk 10 above)
Comment (dspinellis) [reply epidose/device/update_filter_d.sh]: Yes, you can start with one refactoring commit, which doesn't add/change anything

File: epidose/device/wps_scanner_d.sh (hunk 12 above)
Comment (dspinellis) [reply epidose/device/wps_scanner_d.sh]: Seems right.  In general, you should keep in mind such failure modes and ensure they won't occur in production.

File: epidose/device/wps_scanner_d.sh (hunk 14 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: When I was making some tests, I observed that dhclient would return 0 even if not connected to a network and script execution will continue normally. 
See also a [related topic](https://unix.stackexchange.com/questions/8406/dhclient-exits-with-status-0-but-still-no-address).

File: epidose/device/update_filter_d.sh (hunk 10 above)
Comment (stefanos1316) [reply epidose/device/update_filter_d.sh]: Ok, then I will use git rebase to break this comment into two smaller ones.
In the first commit, I will include the migrations of the checks found in the update_filter_d.sh to the util.sh.
In the seconds commit, I will include the WPS-related code.
Ok?

Comment (dspinellis) [reply epidose/device/update_filter_d.sh]: Excellent, looking forward!

File: epidose/device/util.sh [hunk 17]
```diff
@@ -129,6 +136,50 @@ run_python()
   "$PYTHON" "$SCRIPT_DIR"/"$prog".py $DEBUG_FLAG $VERBOSE_FLAG "$@"
 }
 
+# Check the validity of the cuckoo filter
+# preconditions: none 
+# Returns the time required to sleep (in seconds)
```
Comment (dspinellis) [inline epidose/device/util.sh]: I do not see a return value.  You set a global variable, which is confusing. Please rearrange to actually return a value.  It's probably best to output the value on stdout. Document that and modify the function's name to reflect its behavior.

File: epidose/device/util.sh [hunk 18]
```diff
@@ -129,6 +136,50 @@ run_python()
...
+    fi
+  else
+    log "No filter available; download required"
+    mkdir -p "$(dirname $FILTER)"
+  fi	
+}
+
+# Obtain a (new) version of the Cuckoo filter
+# preconditions: none 
+# Returns an exit code that defines
```
Comment (dspinellis) [inline epidose/device/util.sh:167]: Similarly, an exit code has a specific meaning, it's not a global variable.  Use `return` and modify the function's name accordingly to reflect the returned value (`is_...`)

**Review by dspinellis — CHANGES_REQUESTED** [review]
Much clearer; thanks! Only few minor changes are required.

File: epidose/device/util.sh (hunk 17 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: I have set local variables now. Also, I have tried to actually return the values `(return $to_sleep)` and read the value as `to_sleep=$?`, but every time the script was failing if the value was not zero. Therefore, I printed it in the stdout as you said above and created a local variable to read the values (from where the function was invoked). I made also some changes in the documentation. Let me know if this is what you had in mind.

File: epidose/device/update_filter_d.sh [hunk 19]
```diff
@@ -64,17 +56,15 @@ get_filter()
...
     wifi_acquire
-    if err=$(curl --silent --show-error --fail --output "$FILTER.new" \
-      "$SERVER_URL/filter" 2>&1) ; then
-      wifi_release
-      # Atomically replace existing filter with new one
-      mv "$FILTER.new" "$FILTER"
-      log "New filter obtained: $(stat -c %s "$FILTER") bytes"
+
+    # Tries to get a new cuckoo filter from the ha-server
+    local exit_code=$(is_trying_to_get_new_filter)
```
Comment (dspinellis) [inline epidose/device/update_filter_d.sh]: This is getting the output, not the exit code.

File: epidose/device/util.sh [hunk 20]
```diff
@@ -129,6 +136,54 @@ run_python()
   "$PYTHON" "$SCRIPT_DIR"/"$prog".py $DEBUG_FLAG $VERBOSE_FLAG "$@"
 }
 
+# Check the validity of the cuckoo filter
+# preconditions: none 
+# Returns, through the stdout,
```
Comment (dspinellis) [inline epidose/device/util.sh]: Outputs, not returns.

File: epidose/device/util.sh [hunk 21]
```diff
@@ -129,6 +136,54 @@ run_python()
...
+# the time until a Cuckoo filter is valid (in seconds)
```
Comment (dspinellis) [inline epidose/device/util.sh]: time interval, not time

File: epidose/device/util.sh [hunk 22]
```diff
@@ -129,6 +136,54 @@ run_python()
...
+# If 0 is returned, the Cuckoo filter is stale.
```
Comment (dspinellis) [inline epidose/device/util.sh:143]: An output value of 0 means that the Cuckoo filter is stale, right?

File: epidose/device/util.sh [hunk 23]
```diff
@@ -129,6 +136,54 @@ run_python()
...
+# If 1 is returned, then cuckoo filter was not obtained.
+is_trying_to_get_new_filter()
+{
+  if err=$(curl --silent --show-error --fail --output "$FILTER.new" \
+    "$SERVER_URL/filter" 2>&1) ; then
+    # Atomically replace existing filter with new one
+    mv "$FILTER.new" "$FILTER"
+    log "New filter obtained: $(stat -c %s "$FILTER") bytes"
+    local exit_code=0
+    echo "$exit_code"
```
Comment (dspinellis) [inline epidose/device/util.sh]: `echo 0` would be enough, but shouldn't you be using `return`?

File: epidose/device/util.sh [hunk 24]
```diff
@@ -129,6 +136,54 @@ run_python()
...
+  else
+    log "Unable to get filter: $err"
+    local exit_code=1
+    echo "$exit_code"
```
Comment (dspinellis) [inline epidose/device/util.sh]: Ditto

**Review by dspinellis — CHANGES_REQUESTED** [review]

**Comment by dspinellis** [comment]
Unfortunately, more changes are needed.  Please ensure that the shell scripts you modify pass cleanly through *shellcheck* before pushing.

File: epidose/device/util.sh (hunk 20 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: Done.

File: epidose/device/util.sh (hunk 21 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: Done.

File: epidose/device/util.sh (hunk 22 above)
Comment (stefanos1316) [reply epidose/device/util.sh:143]: Yes.

File: epidose/device/update_filter_d.sh (hunk 19 above)
Comment (stefanos1316) [reply epidose/device/update_filter_d.sh]: So I guess here something like this is needed, right?

```
is_try_to_get_new_filter
local exit_code=$?
```

File: epidose/device/util.sh (hunk 23 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: I realized today that I was confused with the return values and return exit code for the BASH scripting.
Done.

File: epidose/device/util.sh (hunk 24 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: Ok done.
I will test them and submit a squashed commit.
Regarding the shellcheck, I thought it was running as a pre-commit hook, no?

Comment (stefanos1316) [reply epidose/device/util.sh]: Ok done.

black................................................(no files to check)Skipped
flake8...............................................(no files to check)Skipped
Test shell scripts with shellcheck.......................................Passed

File: epidose/device/util.sh [hunk 25]
```diff
@@ -129,6 +136,53 @@ run_python()
...
+    fi
+  else
+    log "No filter available; download required"
+    mkdir -p "$(dirname $FILTER)"
+  fi	
+}
+
+# Obtain a (new) version of the Cuckoo filter
+# preconditions: none 
+# Returns, through the stdout, an exit code that defines
```
Comment (dspinellis) [inline epidose/device/util.sh]: It returns the exit code, not a value through stdout.

File: epidose/device/util.sh [hunk 26]
```diff
@@ -129,6 +136,53 @@ run_python()
...
+# whether a request to the ha-sever, to fetch a new cuckoo filter,
+# was successful.
+# If 1 is returned, then cuckoo filter was not obtained.
+is_trying_to_get_new_filter()
```
Comment (dspinellis) [inline epidose/device/util.sh]: The name should be `get_new_filter`.  No `is`, no `trying`.

File: epidose/device/util.sh [hunk 27]
```diff
@@ -129,6 +136,53 @@ run_python()
...
+{
+  if err=$(curl --silent --show-error --fail --output "$FILTER.new" \
```
Comment (dspinellis) [inline epidose/device/util.sh:173]: You want curl's exit code here, not its stdout.  So `if curl ...`.  I can't see how this could have passed a test.

**Review by dspinellis — CHANGES_REQUESTED** [review]
Thank for providing a single commit merged commit!  Not there yet, but  inclosing…

File: epidose/device/util.sh (hunk 25 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: Done

File: epidose/device/util.sh (hunk 26 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: I have named it likes this because update_filter_d.sh has a functions name get_filter which invokes the above function,
but I will name it as you suggested.

File: epidose/device/util.sh (hunk 27 above)
Comment (stefanos1316) [reply epidose/device/util.sh:173]: Which test do you refer to?
Note that I haven't modified this part of code.
I found it like this from the initial version of the [update_filter_d.sh](https://github.com/eellak/epidose/blob/master/epidose/device/update_filter_d.sh).

File: epidose/device/util.sh (hunk 26 above)
Comment (dspinellis) [reply epidose/device/util.sh]: Think carefully: what is the difference between these two functions?  Can both be visibible?  Is so, what is the most appropriate name for each?

File: epidose/device/util.sh (hunk 27 above)
Comment (stefanos1316) [reply epidose/device/util.sh:173]: Don't we need the err messages because in case of a failure we might have a clear output for the following` log "Unable to get filter: $err"`, no?

File: epidose/device/util.sh (hunk 26 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: I think I found a better way to describe this.
I have also submitted the new changes, please let me know what you think.

File: epidose/device/util.sh [hunk 28]
```diff
@@ -129,6 +136,53 @@ run_python()
...
+      echo "$validity_time"
+    fi
+  else
+    log "No filter available; download required"
+    mkdir -p "$(dirname $FILTER)"
+  fi	
+}
+
+# Obtain a (new) version of the Cuckoo filter
+# preconditions: none
```
Comment (dspinellis) [inline epidose/device/util.sh]: I think the precondition here is that WiFi should have been turned on, no?

File: epidose/device/util.sh [hunk 29]
```diff
@@ -129,6 +136,53 @@ run_python()
...
+
+# Obtain a (new) version of the Cuckoo filter
+# preconditions: none 
+# Returns an exit code that defines
+# whether a request to the ha-sever, to fetch a new cuckoo filter,
+# was successful.
+# If 1 is returned, then cuckoo filter was not obtained.
+get_new_filter()
+{
+  if curl --silent --show-error --fail --output "$FILTER.new" \
```
Comment (dspinellis) [inline epidose/device/util.sh]: Please use the previours `err=$(..)` construct to obtain and log curl's output.

**Review by dspinellis — COMMENTED** [review]
Much much clearer now.  I asked for only two changes.  Please verify no other preconditions / postconditions are missed related to WiFi state.

File: epidose/device/util.sh (hunk 28 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: Right, for the current two functions that I am adding only the second one needs WiFi connection.

File: epidose/device/util.sh (hunk 29 above)
Comment (stefanos1316) [reply epidose/device/util.sh]: Ok, done.
I have updated my commit.

**Comment by dspinellis** [comment]
Thank you!  I merged it.  Please implement the following change:
```
-    if ! get_new_filter; then
+    if get_new_filter; then
       wifi_release
-      log "Will retry in $WIFI_RETRY_TIME s"
-      sleep "$WIFI_RETRY_TIME"
+      break
     else
       wifi_release
-      break
-    fi
+      log "Will retry in $WIFI_RETRY_TIME s"
+      sleep "$WIFI_RETRY_TIME"
+    fi
```

and ensure the script passes *shellcheck*.
```
Test shell scripts with shellcheck.......................................Failed
- hook id: shellcheck
- exit code: 1

In epidose/device/update_filter_d.sh line 27:
MAX_FILTER_AGE=$((6 * 60 * 60))
^-- SC2034: MAX_FILTER_AGE appears unused. Verify use (or export if used externally).
```
